power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/sparse_lu_solver.hpp

```diff
@@ -328,6 +405,210 @@ template <class Tensor, class RHSVector, class XVector> class SparseLUSolver {
     std::shared_ptr<IdxVector const> diag_lu_;
+    // cache value for pivot perturbation for the factorize step
+    bool has_pivot_perturbation_{false};
+    double matrix_norm_{};
+    std::optional<std::vector<Tensor>> original_matrix_;
+    // cache value for iterative refinement for the solve step
+    std::optional<std::vector<XVector>> dx_;
+    std::optional<std::vector<RHSVector>> residual_;
+    std::optional<std::vector<RHSVector>> rhs_;
+
+    void solve_with_refinement(std::vector<Tensor> const& data,        // pre-factoirzed data, const ref
+                               BlockPermArray const& block_perm_array, // pre-calculated permutation, const ref
+                               std::vector<RHSVector> const& rhs, std::vector<XVector>& x) {
+        // initialize refinement
+        initialize_refinement(rhs, x);
+        double backward_error{10.0};
+        solve_once(data, block_perm_array, rhs_.value(), x);
+        Idx num_iter{};
+        // iterate until convergence
+        while (backward_error > epsilon_perturbation) {
+            if (num_iter++ == max_iterative_refinement) {
+                throw SparseMatrixError{};
+            }
+            // calculate residual
+            calculate_residual(x);
+            // solve with residual
+            solve_once(data, block_perm_array, residual_.value(), dx_.value());
+            // calculate backward error and then iterate x
+            backward_error = iterate_and_backward_error(x);
+        }
+        // reset refinement cache
+        reset_refinement_cache();
+    }
+
+    void reset_refinement_cache() {
+        dx_.reset();
+        residual_.reset();
+        rhs_.reset();
+    }
+
+    void initialize_refinement(std::vector<RHSVector> const& rhs, std::vector<XVector>& x) {
+        // save a copy of rhs
+        rhs_ = rhs;
+        // initialize x to zero
+        for (Idx row = 0; row != size_; ++row) {
+            x[row] = XVector{};
+        }
+        // initialize residual to rhs
+        // because r = b - A * x = b - 0 = b
+        residual_ = rhs_;
+        // initialize dx to zero
+        // pre-allocate memory
+        dx_ = x;
+    }
+
+    void calculate_residual(std::vector<XVector> const& x) {
+        auto const& row_indptr = *row_indptr_;
+        auto const& col_indices = *col_indices_;
+        auto const& original_matrix = original_matrix_.value();
+        auto const& rhs = rhs_.value();
+        auto& residual = residual_.value();
+        // calculate residual
+        for (Idx row = 0; row != size_; ++row) {
+            residual[row] = rhs[row];
+            // loop all columns
+            for (Idx idx = row_indptr[row]; idx != row_indptr[row + 1]; ++idx) {
+                // subtract
+                residual[row] -= dot(original_matrix[idx], x[col_indices[idx]]);
+            }
+        }
+    }
+
+    double iterate_and_backward_error(std::vector<XVector>& x) {
+        double backward_error_max_demoniator = 0.0;
```

Comment:
typo
```suggestion
        double backward_error_max_denominator = 0.0;
```
unless we're summoning demons here, in which case: let's keep it in 😄

Reply:
Will not be relevant when implementing the new approach